packages/analytics/amplify_analytics_pinpoint_dart/lib/src/impl/analytics_client/event_client/string_database/index_db/index_db_string_db.dart

```diff
@@ -0,0 +1,241 @@
+
+  /// The name of the database
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#name
+  String get databaseName => 'amplify_analytics_pinpoint';
+
+  /// The name of the object store
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#object-store-name
+  final storeName = 'analytics_cached_events';
+
+  late final Future<void> _databaseOpenEvent = _openDatabase();
+
+  late final IDBDatabase _database;
+
+  /// Checks for IDB availability and attempts to open the database.
+  Future<void> _openDatabase() async {
+    if (indexedDB == null) {
+      throw const InvalidStateException('IndexDB is not available');
+    }
+    final openRequest = indexedDB!.open(databaseName, 1)
+      ..onupgradeneeded = (event) {
+        final database = event.target.result;
+        final objectStoreNames = database.objectStoreNames;
+        if (!objectStoreNames.contains(storeName)) {
+          database.createObjectStore(
+            storeName,
+            keyPath: 'id',
+            autoIncrement: true,
+          );
+        }
+      };
+    _database = await openRequest.future;
+  }
+
+  /// Returns a new [IDBObjectStore] instance after waiting for initialization
+  /// to complete.
+  IDBObjectStore _getObjectStore() {
+    final transaction = _database.transaction(
+      storeName,
+      mode: IDBTransactionMode.readwrite,
+    );
+    final store = transaction.objectStore(storeName);
+    return store;
+  }
+
+  @override
+  FutureOr<void> addString(String string) async {
+    await _databaseOpenEvent;
+    await _getObjectStore()
+        .push(StringDatabaseElement.createEntry(string))
+        .future;
+  }
+
+  Future<List<StringDatabaseElement>> getAllStrings({int? maxToGet}) async {
+    final readValues = <StringDatabaseElement>[];
+
+    await _databaseOpenEvent;
+    final store = _getObjectStore();
+    final request = store.getAll(null, maxToGet);
+
+    await request.future;
+
+    for (final elem in request.result) {
+      final value = elem as Object;
+      final id = getProperty<int>(value, 'id');
+      final string = getProperty<String>(value, 'string');
+      readValues.add(
+        StringDatabaseElement(id: id, string: string),
+      );
+    }
+    return readValues;
+  }
+
+  @override
+  Future<List<StringDatabaseElement>> getStrings(int maxToGet) async {
+    return getAllStrings(maxToGet: maxToGet);
+  }
+
+  @override
+  FutureOr<void> deleteStrings(Iterable<int> idsToDelete) async {
+    if (idsToDelete.isEmpty) return;
+
+    await _databaseOpenEvent;
+    final store = _getObjectStore();
+
+    var start = idsToDelete.first;
+    var previous = idsToDelete.first;
+
+    for (final id in idsToDelete) {
+      if (previous != id - 1) {
+        store.deleteByKeyRange(
+          IDBKeyRange.bound(start, previous),
+        );
+        start = id;
+      }
+      previous = id;
+    }
+    await store
+        .deleteByKeyRange(IDBKeyRange.bound(start, idsToDelete.last))
+        .future;
```

Comment:
This logic makes no sense to me. If `idsToDelete = [1, 2, 3, 4, 5]`, this will result in the following calls:

```
deleteByKeyRange (1, 1)
deleteByKeyRange (1, 5)
```

I had to work this out on paper it's very confusing. Please clean up and add tests for this logic.

Reply:
@dnys1 I think the main issue was I missed an edge case.  So updated code to: 

    var previous = idsToDelete.first - 1;

I've added comments explaining how the logic works. 

This is how it works basically: 

We want to delete sequential id ranges.  We keep track of the start index of the range.  We iterate until the next id is no longer sequential. We then delete that range and continue.

Reply:
I've added a lot of extra tests to make sure it's right.